### PR TITLE
build: C++20 migration (again^3)

### PR DESCRIPTION
**Pull request by scarf005**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [x] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [x] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

### Optional

- [x] This is a PR that modifies build process or code organization.
  - [x] Please document the changes in the appropriate location in the `doc/` folder.
  - [x] If documentation for this feature or process does not exist, please write it.
  - [x] If the change alters versions of software required to build or work with the game, please document it.

## Purpose of change

continuation of #1624 and #2003 and #2341

- applied `<=>` operator and concepts to showcase new C++20 features
- use sweet sweet [ranges](https://en.cppreference.com/w/cpp/ranges)

## Describe the solution

- reapplied commits from #2341
- disabled some filesystem tests that were constantly failing on windows (refer to [`10d6c53` (#5514)](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/5514/commits/10d6c53b5fb8b05e4782438cccf5d81ed462c7f5))

## Describe alternatives you've considered

completely overhaul codebase to use `std::filesystem`? MSVC tests results are really weird, fails even on ascii filesystem tests and not sure why

## Testing

intentionally ignoring clang-tidy failures because 
- the automated changes are too large
- currently preparing for following PR that applies all available automated fixes

### CODE REVIEW

**Review by OrenAudeles — COMMENTED** [review]
Looks fine, just have some questions. The effect the Arithmetic concept has on the templates in `units_def.h` is crazy!

File: src/vehicle_use.cpp [hunk 1]
```diff
@@ -137,7 +137,7 @@ void vehicle::add_toggle_to_opts( std::vector<uilist_entry> &options,
                               name );
     options.emplace_back( -1, allow, key, msg );
 
-    actions.emplace_back( [ = ] {
+    actions.emplace_back( [ =, this ] {
```
[the remark this answers is not in the document]
Comment (scarf005) [reply src/vehicle_use.cpp:140]: since c++20 implicit this caputure has been made error

[the remark this answers is not in the document]
Comment (scarf005) [reply src/vehicle_use.cpp:140]: since c++20 implicit this caputure has been made error

[the remark this answers is not in the document]
Comment (OrenAudeles) [reply src/vehicle_use.cpp:140]: okay coolie :+1:

**Comment by OrenAudeles** [comment]
the `Clang-tidy (tiles) / build (pull_request)` failure is emitting this error. Not sure what to do about that one since it's in sol files.
Screenshot from 2024-10-06 02-12-25
and the `Cataclysm Windows build (Cmake + MSVC) / Build (pull_request)` failure is in `tests/filesystem_test.cpp` at the `filesystem_utf8[_comp|_decomp]` tests. `filesystem_ascii` passes just fine, so something is still weird about filesystem tests. The failure point is all on the same line checking for directory existence right after calling `bool remove_tree( const std::string &path )` so it's likely failing there.

**Comment by chaosvolt** [comment]
Oren asked me in the discord server to uncomment the lines that were causing problems and test if it works on mine or not, it indeed does fail as expected on windows for me:
image

File: src/sol/sol.hpp [hunk 2]
```diff
@@ -38,7 +38,7 @@ static_assert(false, "sol.hpp must be included via catalua_sol.h");
 
 // beginning of sol/version.hpp
 
-#include <sol/config.hpp>
+#include "./config.hpp"
```
Comment (OrenAudeles) [inline src/sol/sol.hpp:41]: Similarly to `forward.hpp` the changes in this file are mostly cosmetic whitespace removals that don't change functionality. Could change back to `#include <sol/config.hpp>` while getting rid of all changes to this file and clang-tidy will probably shut up about it.

File: android/app/src/main/java/org/libsdl/app/SDL.java [hunk 3]
```diff
@@ -42,42 +42,34 @@ public static void loadLibrary(String libraryName) throws UnsatisfiedLinkError,
...
-                System.loadLibrary(libraryName);
-            }
-            catch (final UnsatisfiedLinkError ule) {
-                throw ule;
-            }
-            catch (final SecurityException se) {
-                throw se;
-            }
-        }        
+            System.loadLibrary(libraryName);
```
Comment (OrenAudeles) [inline android/app/src/main/java/org/libsdl/app/SDL.java:71]: Is removing the try-catch block here safe to do?

**Review by OrenAudeles — APPROVED** [review]
Other than clang-tidy whining about a static assertion error that doesn't come up while actually building, and claims of not being able to find `<lua.h>`, both involving `sol` I don't see anything else preventing merging. The clang-tidy errors are nonsense.

Will need to look at the filesystem_test issues revolving around `remove_tree` and `dir_exists` since they only show up as an issue on the windows build. Thankfully `remove_tree` is only used as part of this one test and not the game proper.

File: src/iuse.cpp [hunk 4]
```diff
@@ -363,9 +363,31 @@ static constexpr int RADIO_PER_TURN = 25;
 
 #include "iuse_software.h"
 
-struct extended_photo_def;
+
 struct object_names_collection;
 
+struct extended_photo_def : public JsonDeserializer, public JsonSerializer {
```
[the remark this answers is not in the document]
Comment (scarf005) [reply src/iuse.cpp:369]: iirc it was somewhat related to build error but i forgot

**Review by chaosvolt — APPROVED** [review]
1. Compiled and loaded a save.
2. Saved game and deleted world without issue.
3. Created a game without issue.
4. Started new game in new world, chargen and entering game world also works fine.
5. Saved this new game for good measure too.

**Pull request merged**